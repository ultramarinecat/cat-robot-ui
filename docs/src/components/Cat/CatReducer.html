<!DOCTYPE html>

<html>
<head>
  <title>CatReducer.js</title>
  <meta http-equiv="content-type" content="text/html; charset=UTF-8">
  <meta name="viewport" content="width=device-width, target-densitydpi=160dpi, initial-scale=1.0; maximum-scale=1.0; user-scalable=0;">
  <link rel="stylesheet" media="all" href="../../../docco.css" />
</head>
<body>
  <div id="container">
    <div id="background"></div>
    
      <ul id="jump_to">
        <li>
          <a class="large" href="javascript:void(0);">Jump To &hellip;</a>
          <a class="small" href="javascript:void(0);">+</a>
          <div id="jump_wrapper">
          <div id="jump_page_wrapper">
            <div id="jump_page">
              
                
                <a class="source" href="../App/App.html">
                  src/components/App/App.js
                </a>
              
                
                <a class="source" href="../App/ErrorBoundary.html">
                  src/components/App/ErrorBoundary.js
                </a>
              
                
                <a class="source" href="../App/i18n.html">
                  src/components/App/i18n.js
                </a>
              
                
                <a class="source" href="../Auth/Auth.html">
                  src/components/Auth/Auth.js
                </a>
              
                
                <a class="source" href="../Auth/AuthActions.html">
                  src/components/Auth/AuthActions.js
                </a>
              
                
                <a class="source" href="../Auth/AuthReducer.html">
                  src/components/Auth/AuthReducer.js
                </a>
              
                
                <a class="source" href="../Button/Button.html">
                  src/components/Button/Button.js
                </a>
              
                
                <a class="source" href="Cat.html">
                  src/components/Cat/Cat.js
                </a>
              
                
                <a class="source" href="CatActions.html">
                  src/components/Cat/CatActions.js
                </a>
              
                
                <a class="source" href="CatReducer.html">
                  src/components/Cat/CatReducer.js
                </a>
              
                
                <a class="source" href="fulltilt.html">
                  src/components/Cat/fulltilt.js
                </a>
              
                
                <a class="source" href="../Connection/Connection.html">
                  src/components/Connection/Connection.js
                </a>
              
                
                <a class="source" href="../Connection/ConnectionActions.html">
                  src/components/Connection/ConnectionActions.js
                </a>
              
                
                <a class="source" href="../Connection/ConnectionReducer.html">
                  src/components/Connection/ConnectionReducer.js
                </a>
              
                
                <a class="source" href="../Connection/messages.html">
                  src/components/Connection/messages.js
                </a>
              
                
                <a class="source" href="../Intro/Intro.html">
                  src/components/Intro/Intro.js
                </a>
              
                
                <a class="source" href="../Intro/IntroActions.html">
                  src/components/Intro/IntroActions.js
                </a>
              
                
                <a class="source" href="../Particles/Particles.html">
                  src/components/Particles/Particles.js
                </a>
              
                
                <a class="source" href="../Pulse/Pulse.html">
                  src/components/Pulse/Pulse.js
                </a>
              
                
                <a class="source" href="../Router/Router.html">
                  src/components/Router/Router.js
                </a>
              
                
                <a class="source" href="../Sound/CatSound.html">
                  src/components/Sound/CatSound.js
                </a>
              
                
                <a class="source" href="../Sound/Mute.html">
                  src/components/Sound/Mute.js
                </a>
              
                
                <a class="source" href="../Sound/Sound.html">
                  src/components/Sound/Sound.js
                </a>
              
                
                <a class="source" href="../Sound/TurningSound.html">
                  src/components/Sound/TurningSound.js
                </a>
              
                
                <a class="source" href="../Spinner/Spinner.html">
                  src/components/Spinner/Spinner.js
                </a>
              
                
                <a class="source" href="../Typing/Typing.html">
                  src/components/Typing/Typing.js
                </a>
              
                
                <a class="source" href="../../configureStore.html">
                  src/configureStore.js
                </a>
              
                
                <a class="source" href="../../index.html">
                  src/index.js
                </a>
              
                
                <a class="source" href="../../serviceWorker.html">
                  src/serviceWorker.js
                </a>
              
                
                <a class="source" href="../../setupTests.html">
                  src/setupTests.js
                </a>
              
            </div>
          </div>
        </li>
      </ul>
    
    <ul class="sections">
        
          <li id="title">
              <div class="annotation">
                  <h1>CatReducer.js</h1>
              </div>
          </li>
        
        
        
        <li id="section-1">
            <div class="annotation">
              
              <div class="pilwrap ">
                <a class="pilcrow" href="#section-1">&#182;</a>
              </div>
              <p>Cat reducer</p>

            </div>
            
            <div class="content"><div class='highlight'><pre>
<span class="hljs-keyword">import</span> {
  CAT_DETECTED,
  ROBOT_CURRENTLY_TURNING,
  ROBOT_TURNING_LEFT,
  ROBOT_TURNING_RIGHT,
  ROBOT_COMPLETED_TURN,
  CONNECTION_SUCCESSFUL,
  CLEAR_CAT_DETECTED,
  DISCONNECTED
} <span class="hljs-keyword">from</span> <span class="hljs-string">'../Connection/ConnectionActions.js'</span>;

<span class="hljs-keyword">export</span> <span class="hljs-keyword">default</span> <span class="hljs-function"><span class="hljs-keyword">function</span> <span class="hljs-title">connectionReducer</span>(<span class="hljs-params">state = {}, action</span>) </span>{
  <span class="hljs-keyword">switch</span> (action.type) {
    <span class="hljs-keyword">case</span> ROBOT_CURRENTLY_TURNING:
      <span class="hljs-keyword">return</span> {
        ...state,
        ...{
          <span class="hljs-attr">robotCurrentlyTurning</span>: <span class="hljs-literal">true</span>,
          <span class="hljs-attr">robotTurningLeft</span>: <span class="hljs-literal">false</span>,
          <span class="hljs-attr">robotTurningRight</span>: <span class="hljs-literal">false</span>,
          <span class="hljs-attr">robotCompletedTurn</span>: <span class="hljs-literal">false</span>,
          <span class="hljs-attr">catDetected</span>: <span class="hljs-literal">false</span>
        }
      };
    <span class="hljs-keyword">case</span> ROBOT_TURNING_LEFT:
      <span class="hljs-keyword">return</span> {
        ...state,
        ...{
          <span class="hljs-attr">robotTurningLeft</span>: <span class="hljs-literal">true</span>,
          <span class="hljs-attr">robotTurningRight</span>: <span class="hljs-literal">false</span>,
          <span class="hljs-attr">robotCompletedTurn</span>: <span class="hljs-literal">false</span>,
          <span class="hljs-attr">robotCurrentlyTurning</span>: <span class="hljs-literal">false</span>,
          <span class="hljs-attr">catDetected</span>: <span class="hljs-literal">false</span>
        }
      };
    <span class="hljs-keyword">case</span> ROBOT_TURNING_RIGHT:
      <span class="hljs-keyword">return</span> {
        ...state,
        ...{
          <span class="hljs-attr">robotTurningRight</span>: <span class="hljs-literal">true</span>,
          <span class="hljs-attr">robotTurningLeft</span>: <span class="hljs-literal">false</span>,
          <span class="hljs-attr">robotCompletedTurn</span>: <span class="hljs-literal">false</span>,
          <span class="hljs-attr">robotCurrentlyTurning</span>: <span class="hljs-literal">false</span>,
          <span class="hljs-attr">catDetected</span>: <span class="hljs-literal">false</span>
        }
      };
    <span class="hljs-keyword">case</span> ROBOT_COMPLETED_TURN:
      <span class="hljs-keyword">return</span> {
        ...state,
        ...{
          <span class="hljs-attr">robotCompletedTurn</span>: <span class="hljs-literal">true</span>,
          <span class="hljs-attr">robotTurningLeft</span>: <span class="hljs-literal">false</span>,
          <span class="hljs-attr">robotTurningRight</span>: <span class="hljs-literal">false</span>,
          <span class="hljs-attr">robotCurrentlyTurning</span>: <span class="hljs-literal">false</span>,
          <span class="hljs-attr">catDetected</span>: <span class="hljs-literal">false</span>
        }
      };
    <span class="hljs-keyword">case</span> CAT_DETECTED:
      <span class="hljs-keyword">return</span> {
        ...state,
        ...{
          <span class="hljs-attr">catDetected</span>: <span class="hljs-literal">true</span>,
          <span class="hljs-attr">robotTurningLeft</span>: <span class="hljs-literal">false</span>,
          <span class="hljs-attr">robotTurningRight</span>: <span class="hljs-literal">false</span>,
          <span class="hljs-attr">robotCompletedTurn</span>: <span class="hljs-literal">false</span>,
          <span class="hljs-attr">robotCurrentlyTurning</span>: <span class="hljs-literal">false</span>
        }
      };
    <span class="hljs-keyword">case</span> CLEAR_CAT_DETECTED:
      <span class="hljs-keyword">return</span> {
        ...state,
        ...{
          <span class="hljs-attr">catDetected</span>: <span class="hljs-literal">false</span>
        }
      };
    <span class="hljs-keyword">case</span> CONNECTION_SUCCESSFUL:
      <span class="hljs-keyword">return</span> {
        ...state,
        ...{
          <span class="hljs-attr">disconnected</span>: <span class="hljs-literal">false</span>
        }
      };
    <span class="hljs-keyword">case</span> DISCONNECTED:
      <span class="hljs-keyword">return</span> {
        ...state,
        ...{
          <span class="hljs-attr">disconnected</span>: <span class="hljs-literal">true</span>
        }
      };
    <span class="hljs-keyword">default</span>:
      <span class="hljs-keyword">return</span> state;
  }
}</pre></div></div>
            
        </li>
        
    </ul>
  </div>
</body>
</html>
