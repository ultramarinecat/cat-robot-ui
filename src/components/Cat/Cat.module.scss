@charset 'utf-8';

////
//// Cat styles
////

@import 'base.scss';

.main {
  max-width: rem(600);
  position: relative;
  @extend %center-block;
}
