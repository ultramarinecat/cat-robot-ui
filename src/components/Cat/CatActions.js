//
// Cat actions
//

export const LEFT_TILT = 'LEFT_TILT';
export const RIGHT_TILT = 'RIGHT_TILT';

export function leftTilt() {
  return {
    type: LEFT_TILT
  };
}

export function rightTilt() {
  return {
    type: RIGHT_TILT
  };
}
