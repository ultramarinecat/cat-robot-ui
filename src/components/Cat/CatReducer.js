//
// Cat reducer
//

import {
  CAT_DETECTED,
  ROBOT_CURRENTLY_TURNING,
  ROBOT_TURNING_LEFT,
  ROBOT_TURNING_RIGHT,
  ROBOT_COMPLETED_TURN,
  CONNECTION_SUCCESSFUL,
  CLEAR_CAT_DETECTED,
  DISCONNECTED
} from '../Connection/ConnectionActions.js';

export default function connectionReducer(state = {}, action) {
  switch (action.type) {
    case ROBOT_CURRENTLY_TURNING:
      return {
        ...state,
        ...{
          robotCurrentlyTurning: true,
          robotTurningLeft: false,
          robotTurningRight: false,
          robotCompletedTurn: false,
          catDetected: false
        }
      };
    case ROBOT_TURNING_LEFT:
      return {
        ...state,
        ...{
          robotTurningLeft: true,
          robotTurningRight: false,
          robotCompletedTurn: false,
          robotCurrentlyTurning: false,
          catDetected: false
        }
      };
    case ROBOT_TURNING_RIGHT:
      return {
        ...state,
        ...{
          robotTurningRight: true,
          robotTurningLeft: false,
          robotCompletedTurn: false,
          robotCurrentlyTurning: false,
          catDetected: false
        }
      };
    case ROBOT_COMPLETED_TURN:
      return {
        ...state,
        ...{
          robotCompletedTurn: true,
          robotTurningLeft: false,
          robotTurningRight: false,
          robotCurrentlyTurning: false,
          catDetected: false
        }
      };
    case CAT_DETECTED:
      return {
        ...state,
        ...{
          catDetected: true,
          robotTurningLeft: false,
          robotTurningRight: false,
          robotCompletedTurn: false,
          robotCurrentlyTurning: false
        }
      };
    case CLEAR_CAT_DETECTED:
      return {
        ...state,
        ...{
          catDetected: false
        }
      };
    case CONNECTION_SUCCESSFUL:
      return {
        ...state,
        ...{
          disconnected: false
        }
      };
    case DISCONNECTED:
      return {
        ...state,
        ...{
          disconnected: true
        }
      };
    default:
      return state;
  }
}
