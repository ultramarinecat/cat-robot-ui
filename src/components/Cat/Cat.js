//
// Cat component
// Manages interaction with cat robot by sending messages based on device orientation events
// and displays updates
//

import React, { memo, useState, useEffect, useReducer } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';

import { withTranslation } from 'react-i18next';
import log from 'loglevel';

import GyroNorm from 'gyronorm';
import './fulltilt.js';

import Pulse from '../Pulse/Pulse.js';
import Sound from '../Sound/Sound.js';
import Typing from '../Typing/Typing.js';

import { leftTilt, rightTilt } from './CatActions.js';

import styles from './Cat.module.scss';

export const REGISTER_TILT_ANGLE = 20;
export const TURN_REQUEST_MIN_INTERVAL = 5000;
const ORIENTATION_EVENT_FREQUENCY = 300;

const PLAY_CAT_SOUND = 'PLAY_CAT_SOUND';
const PLAY_TURNING_SOUND = 'PLAY_TURNING_SOUND';

const LEFT = 'LEFT';
const RIGHT = 'RIGHT';

let gyronorm;
let timeLastTurnRequest;
let tiltCleared = false;

export function Cat({
  catDetected,
  robotCurrentlyTurning,
  robotTurningLeft,
  robotTurningRight,
  robotCompletedTurn,
  disconnected,
  leftTilt,
  rightTilt,
  t
}) {
  const [{ catSound, turningSound, updateSound }, dispatch] = useReducer(catReducer, {
    catSound: false,
    turningSound: false,
    updateSound: false // update even if sound is the same
  });

  const [status, setStatus] = useState(t('connected'));

  useEffect(() => {
    startListeningForOrientationEvents(leftTilt, rightTilt, dispatch)
      .then(() => {
        log.info('listeining for device orientation events');
      })
      .catch(e => {
        log.warn('a problem occured initializing gyronorm', e);
        setStatus(t('device_not_supported'));
      });

    return stopListeningForOrientationEvents;
  }, []);

  useEffect(() => {
    if (disconnected) {
      stopListeningForOrientationEvents();
      setStatus(t('disconnected'));
      return;
    }

    if (catDetected) {
      setStatus(t('cat_detected'));
      // play cat sound when cat is detected!
      dispatch({ type: PLAY_CAT_SOUND });
      return;
    }

    if (robotCurrentlyTurning) {
      setStatus(t('currently_turning'));
      return;
    }

    if (robotTurningLeft) {
      setStatus(t('turning_left'));
      return;
    }

    if (robotTurningRight) {
      setStatus(t('turning_right'));
      return;
    }

    if (robotCompletedTurn) {
      setStatus(t('turn_completed'));
    }
  }, [
    catDetected,
    robotCurrentlyTurning,
    robotTurningLeft,
    robotTurningRight,
    robotCompletedTurn,
    disconnected
  ]);

  return (
    <div className={styles.main}>
      <Pulse />
      <Typing text={status} datatest-id="status" />

      <Sound
        playCatSound={catSound}
        playTurningSound={turningSound}
        updateSound={updateSound}
        data-testid="sound"
      />
    </div>
  );
}

// start listening for orientation events
function startListeningForOrientationEvents(leftTilt, rightTilt, dispatch) {
  gyronorm = new GyroNorm();
  const options = {
    frequency: ORIENTATION_EVENT_FREQUENCY
  };

  // listen for orientation events
  return gyronorm.init(options).then(() => {
    gyronorm.start(data => {
      // get device orientation around y-axis
      const tiltAngle = data.do.gamma;
      log.trace('tilt angle:', tiltAngle);

      // if angle < min angle to register tilt, clear current tilt
      if (Math.abs(tiltAngle) < REGISTER_TILT_ANGLE) {
        if (!tiltCleared) {
          log.trace('cleared tilt');
          tiltCleared = true;
        }

        return;
      }

      // if angle > min angle to register tilt, and current tilt has been cleared
      if (tiltCleared) {
        // register tilt
        const direction = tiltAngle > 0 ? RIGHT : LEFT;
        registerTilt(direction, leftTilt, rightTilt, dispatch);
        tiltCleared = false;
      }
    });
  });
}

// stop listening for orientation events
function stopListeningForOrientationEvents() {
  if (gyronorm.isRunning()) {
    gyronorm.end();
  }
}

// register tilt, send turn request
function registerTilt(tiltDirection, leftTilt, rightTilt, dispatch) {
  // if time since last turn request < TURN_REQUEST_MIN_INTERVAL, ignore
  const timeTurnRequest = Date.now();

  if (
    timeLastTurnRequest != null &&
    timeTurnRequest - timeLastTurnRequest < TURN_REQUEST_MIN_INTERVAL
  ) {
    log.trace(`skipping ${tiltDirection} tilt`);
    return;
  }

  // note time of turn request, play registered turn sound
  timeLastTurnRequest = timeTurnRequest;
  dispatch({ type: PLAY_TURNING_SOUND });

  // send turn request
  if (tiltDirection === LEFT) {
    leftTilt();
  } else {
    rightTilt();
  }
}

function catReducer(state, action) {
  switch (action.type) {
    case PLAY_CAT_SOUND:
      return {
        ...state,
        catSound: true,
        turningSound: false,
        updateSound: !state.updateSound
      };
    case PLAY_TURNING_SOUND:
      return {
        ...state,
        turningSound: true,
        catSound: false,
        updateSound: !state.updateSound
      };
    default:
      return state;
  }
}

Cat.propTypes = {
  catDetected: PropTypes.bool,
  robotCurrentlyTurning: PropTypes.bool,
  robotTurningLeft: PropTypes.bool,
  robotTurningRight: PropTypes.bool,
  robotCompletedTurn: PropTypes.bool,
  disconnected: PropTypes.bool,
  leftTilt: PropTypes.func.isRequired,
  rightTilt: PropTypes.func.isRequired,
  t: PropTypes.func.isRequired
};

const mapStateToProps = state => {
  return {
    catDetected: state.cat.catDetected,
    robotCurrentlyTurning: state.cat.robotCurrentlyTurning,
    robotTurningLeft: state.cat.robotTurningLeft,
    robotTurningRight: state.cat.robotTurningRight,
    robotCompletedTurn: state.cat.robotCompletedTurn,
    disconnected: state.cat.disconnected
  };
};

export default connect(
  mapStateToProps,
  {
    leftTilt,
    rightTilt
  }
)(memo(withTranslation()(Cat)));
