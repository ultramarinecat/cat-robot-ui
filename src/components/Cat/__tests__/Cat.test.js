import React from 'react';
import { render, act, wait } from 'react-testing-library';
import { advanceBy, advanceTo, clear } from 'jest-date-mock';

import GyroNorm from 'gyronorm';
import { Cat, REGISTER_TILT_ANGLE, TURN_REQUEST_MIN_INTERVAL } from '../Cat.js';

const mockPlayCatSound = jest.fn();
const mockPlayTurningSound = jest.fn();

jest.mock('../../Sound/Sound', () => ({ playCatSound, playTurningSound }) => {
  if (playCatSound) {
    mockPlayCatSound();
  }

  if (playTurningSound) {
    mockPlayTurningSound();
  }

  return <div data-testid="sound" />;
});

jest.mock('../../Typing/Typing', () => ({ text }) => (
  <div data-testid="status">{text}</div>
));

jest.mock('../../Pulse/Pulse');
jest.mock('../fulltilt.js');
jest.mock('loglevel');

const CONNECTED = 'connected';
const DISCONNECTED = 'disconnected';
const CAT_DETECTED = 'cat_detected';
const CURRENTLY_TURNING = 'currently_turning';
const TURNING_LEFT = 'turning_left';
const TURNING_RIGHT = 'turning_right';
const TURN_COMPLETED = 'turn_completed';

const STATUS = 'status';
const SOUND = 'sound';

let gyronorm = new GyroNorm();

describe('Cat', () => {
  beforeAll(() => {
    gyronorm = new GyroNorm();
    advanceTo(0);
  });

  afterEach(() => {
    mockPlayCatSound.mockClear();
    mockPlayTurningSound.mockClear();
  });

  afterAll(() => {
    clear();
  });

  it('should display status text and sounds', async () => {
    const { container, getByTestId } = render(
      <Cat leftTilt={jest.fn()} rightTilt={jest.fn()} t={key => key} />
    );

    expect(getByTestId(STATUS)).toBeInTheDocument();
    expect(getByTestId(SOUND)).toBeInTheDocument();

    expect(container.firstChild).toMatchSnapshot();
  });

  it('should call the `leftTilt` prop and play turning sound if device is tilted left', async () => {
    const leftTilt = jest.fn();

    render(<Cat leftTilt={leftTilt} rightTilt={jest.fn()} t={key => key} />);

    await wait(() => {
      expect(gyronorm.start).toHaveBeenCalled();
    });

    act(() => {
      gyronorm.setTiltAngle(0);
      gyronorm.setTiltAngle(-REGISTER_TILT_ANGLE);
    });

    expect(leftTilt).toHaveBeenCalled();
    expect(mockPlayTurningSound).toHaveBeenCalled();
  });

  it('should call the `rightTilt` prop and play turning sound if device is tilted right', async () => {
    const rightTilt = jest.fn();

    render(<Cat leftTilt={jest.fn()} rightTilt={rightTilt} t={key => key} />);

    await wait(() => {
      expect(gyronorm.start).toHaveBeenCalled();
    });

    advanceBy(TURN_REQUEST_MIN_INTERVAL);

    act(() => {
      gyronorm.setTiltAngle(0);
      gyronorm.setTiltAngle(REGISTER_TILT_ANGLE);
    });

    expect(rightTilt).toHaveBeenCalled();
    expect(mockPlayTurningSound).toHaveBeenCalled();
  });

  it(`should not register a tilt if device is tilted less than min angle required to register
      a tilt`, async () => {
    const leftTilt = jest.fn();
    const rightTilt = jest.fn();

    render(<Cat leftTilt={leftTilt} rightTilt={rightTilt} t={key => key} />);

    await wait(() => {
      expect(gyronorm.start).toHaveBeenCalled();
    });

    advanceBy(TURN_REQUEST_MIN_INTERVAL);

    act(() => {
      gyronorm.setTiltAngle(0);
      gyronorm.setTiltAngle(-(REGISTER_TILT_ANGLE - 1));
    });

    advanceBy(TURN_REQUEST_MIN_INTERVAL);

    act(() => {
      gyronorm.setTiltAngle(REGISTER_TILT_ANGLE - 1);
    });

    expect(leftTilt).not.toHaveBeenCalled();
    expect(rightTilt).not.toHaveBeenCalled();
  });

  it(`should not register a tilt if tilt hasn't been reset by tilting device to an angle
      less than min angle required to register tilt`, async () => {
    const rightTilt = jest.fn();
    render(<Cat leftTilt={jest.fn()} rightTilt={rightTilt} t={key => key} />);

    await wait(() => {
      expect(gyronorm.start).toHaveBeenCalled();
    });

    advanceBy(TURN_REQUEST_MIN_INTERVAL);
    act(() => {
      gyronorm.setTiltAngle(REGISTER_TILT_ANGLE + 1);
    });

    expect(rightTilt).toHaveBeenCalled();
    rightTilt.mockClear();

    advanceBy(TURN_REQUEST_MIN_INTERVAL);
    act(() => {
      gyronorm.setTiltAngle(REGISTER_TILT_ANGLE);
    });

    expect(rightTilt).not.toHaveBeenCalled();
    rightTilt.mockClear();

    act(() => {
      gyronorm.setTiltAngle(REGISTER_TILT_ANGLE - 1);
      gyronorm.setTiltAngle(REGISTER_TILT_ANGLE);
    });

    expect(rightTilt).toHaveBeenCalled();
  });

  it(`should not register a tilt if less than min interval has passsed since the previous
      tilt`, async () => {
    const rightTilt = jest.fn();
    render(<Cat leftTilt={jest.fn()} rightTilt={rightTilt} t={key => key} />);

    await wait(() => {
      expect(gyronorm.start).toHaveBeenCalled();
    });

    advanceBy(TURN_REQUEST_MIN_INTERVAL);
    act(() => {
      gyronorm.setTiltAngle(REGISTER_TILT_ANGLE - 1);
      gyronorm.setTiltAngle(REGISTER_TILT_ANGLE);
    });

    expect(rightTilt).toHaveBeenCalled();
    rightTilt.mockClear();

    advanceBy(TURN_REQUEST_MIN_INTERVAL - 1);
    act(() => {
      gyronorm.setTiltAngle(REGISTER_TILT_ANGLE - 1);
      gyronorm.setTiltAngle(REGISTER_TILT_ANGLE);
    });

    expect(rightTilt).not.toHaveBeenCalled();
    rightTilt.mockClear();

    advanceBy(1);
    act(() => {
      gyronorm.setTiltAngle(REGISTER_TILT_ANGLE - 1);
      gyronorm.setTiltAngle(REGISTER_TILT_ANGLE);
    });

    expect(rightTilt).toHaveBeenCalled();
  });

  it(`should initially display that it's connected`, () => {
    const { getByText } = render(
      <Cat leftTilt={jest.fn()} rightTilt={jest.fn()} t={key => key} />
    );

    expect(getByText(CONNECTED)).toBeInTheDocument();
  });

  it(`should display that it's disconnected if \`disconnected\` is true`, async () => {
    const { getByText } = render(
      <Cat
        disconnected={true}
        leftTilt={jest.fn()}
        rightTilt={jest.fn()}
        t={key => key}
      />
    );

    expect(getByText(DISCONNECTED)).toBeInTheDocument();
  });

  it('should display that cat is detected and play cat sound if `catDetected` is true', async () => {
    const { getByText } = render(
      <Cat catDetected={true} leftTilt={jest.fn()} rightTilt={jest.fn()} t={key => key} />
    );

    expect(getByText(CAT_DETECTED)).toBeInTheDocument();
    expect(mockPlayCatSound).toHaveBeenCalled();
  });

  it('should display that robot is currently turning if `robotCurrentlyTurning` is true', async () => {
    const { getByText } = render(
      <Cat
        robotCurrentlyTurning={true}
        leftTilt={jest.fn()}
        rightTilt={jest.fn()}
        t={key => key}
      />
    );

    expect(getByText(CURRENTLY_TURNING)).toBeInTheDocument();
  });

  it('should display that robot is turning left if `robotTurningLeft` is true', async () => {
    const { getByText } = render(
      <Cat
        robotTurningLeft={true}
        leftTilt={jest.fn()}
        rightTilt={jest.fn()}
        t={key => key}
      />
    );

    expect(getByText(TURNING_LEFT)).toBeInTheDocument();
  });

  it('should display that robot is turning right if `robotTurningRight` is true', async () => {
    const { getByText } = render(
      <Cat
        robotTurningRight={true}
        leftTilt={jest.fn()}
        rightTilt={jest.fn()}
        t={key => key}
      />
    );

    expect(getByText(TURNING_RIGHT)).toBeInTheDocument();
  });

  it('should display that robot completed turn if `robotCompletedTurn` is true', async () => {
    const { getByText } = render(
      <Cat
        robotCompletedTurn={true}
        leftTilt={jest.fn()}
        rightTilt={jest.fn()}
        t={key => key}
      />
    );

    expect(getByText(TURN_COMPLETED)).toBeInTheDocument();
  });
});
