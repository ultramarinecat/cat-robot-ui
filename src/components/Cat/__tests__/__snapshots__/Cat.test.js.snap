// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cat should display status text and sounds 1`] = `
<div
  class="main"
>
  <div
    data-testid="status"
  >
    connected
  </div>
  <div
    data-testid="sound"
  />
</div>
`;
